sdk/metric/pipeline.go

```diff
@@ -0,0 +1,136 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.18
+// +build go1.18
+
+package metric // import "go.opentelemetry.io/otel/sdk/metric"
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"go.opentelemetry.io/otel/metric/unit"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+type aggregator interface {
+	Aggregation() metricdata.Aggregation
+}
+type instrumentKey struct {
+	name        string
+	description string
+	unit        unit.Unit
+}
+
+func newPipeline(res *resource.Resource) *pipeline {
+	if res == nil {
+		res = resource.Empty()
+	}
+	return &pipeline{
+		resource:     res,
+		aggregations: make(map[instrumentation.Scope]map[instrumentKey]aggregator),
+	}
+}
+
+// A pipeline connects all of the instruments created by a meter provider to a Reader.
+// This is the object that will be `Reader.register()` when a meter provider is created.
+
+// As instruments are created the instrument should be checked if it exists in the
+// views of a the Reader, and if so each aggregator should be added to the pipeline.
+type pipeline struct {
+	resource *resource.Resource
+
+	sync.Mutex
+	aggregations map[instrumentation.Scope]map[instrumentKey]aggregator
+	callbacks    []func(context.Context)
+}
+
+// addAggregator will stores an aggregator with an instrument description.  This
+// is used when `produce()` is called to create a new metric.
+func (p *pipeline) addAggregator(scope instrumentation.Scope, name, description string, instUnit unit.Unit, agg aggregator) error {
+	p.Lock()
+	defer p.Unlock()
+	if p.aggregations == nil {
+		p.aggregations = map[instrumentation.Scope]map[instrumentKey]aggregator{}
+	}
+	if p.aggregations[scope] == nil {
+		p.aggregations[scope] = map[instrumentKey]aggregator{}
+	}
+	inst := instrumentKey{
+		name:        name,
+		description: description,
+		unit:        instUnit,
+	}
+	if _, ok := p.aggregations[scope][inst]; ok {
+		return fmt.Errorf("instrument already registered: name %s, scope: %s", name, scope)
```

Comment:
Since this is part of a larger construction, I can't quite answer the following: What happens next after this error happens? Is this treated somewhere or returned to the user?

Reply:
@MadVikingGod correct me if I'm wrong, but I imagine this is going to be returned from the instrument creation methods (i.e. https://github.com/open-telemetry/opentelemetry-go/blob/096a1624184285edbf9c82c53ea0b068895915ef/metric/instrument/asyncfloat64/asyncfloat64.go#L24-L34)

Reply:
Yes, this will eventually bubble up to the instrument creation.  This is the first point where an instrument could clash with an already created instrument.

For example, if you have a reader with two views, everything, and rename foo to bar.  The first instrument (foo or bar) that is registered shouldn't have an error, but the second one will.